### ISSUE TITLE
View Model only creating once 

### ISSUE THREAD

**Issue by stezma** [xamarinhq/xamu-infrastructure#10]
I am trying to achieve **View First** MVVM Approach, Suppose If I want to create View Model each time I navigated to the view, What configuration I need to change? 
Right now,

If i navigate from Page A to Page B => View Model will be created for A and B, then I come back and Navigate again to B. Only page B is created the instance of the View Model remains the same. I need to create my view model as well while navigating. How I can achieve that?

**Comment by markjulmar** [comment]
Certainly possible. Can you tell us what your current configuration looks like?

**Comment by stezma** [comment]
In my current application, I am defining my View & View Models in App.xaml.cs  using the below methods

 void RegisterPages()
 {
      var navService = new DependencyServiceWrapper().Get<INavigationService>() as FormsNavigationPageService;
      navService?.RegisterPage(AppPages.DataBindingPage, () => new LoginPage());
      navService?.RegisterPage(AppPages.WebServicePage, () => new UserDetailPage());
 }
 
  private static void RegisterDependencies()
  {
      // Register dependencies.
      DependencyService.Register<LoginViewModel>();
      DependencyService.Register<UserDetailViewModel>();   
  }

And in my ViewModel I am using the below code for navigation.

 await serviceLocator.Get<INavigationService>()
                                        .NavigateAsync(AppPages.LoginnPage);


But in my example, If I navigate to UserDetailPage, it will initialize the corresponding ViewModel during fist navigation. But if I come back and again navigate to the page the view model is live in memory and the constructor is not calling. I would like to create fresh view model each time user navigates to the page. And when the user navigates back to the previous page, I would like to destroy the view model.  How I can achieve  similar feature using our current library. 

In short, I don't want to make all the view models up and running all the time. It should be somewhat dynamic in nature.

**Comment by markjulmar** [comment]
It looks like you are probably grabbing the VM inside each of the `ContentPage` objects and setting it to the `BindingContext` is that correct?

If so, there's an easy fix: just use the `DependencyFetchTarget` parameter of the `Get<T>` method:

```csharp
// In the LoginPage code:
BindingContext = DependencyService.Get<LoginPageViewModel>(DependencyFetchTarget.NewInstance));
```

That parameter indicates to create a new instance of the object. Make sure you don't ever call the method with no parameters, or with the `GlobalInstance` version or it caches it off in a static -- so one instance of the object will never be collected.

**Comment by markjulmar** [comment]
Yes, you could do that in XAML - in fact, I just realized we _do_ have the `FetchTarget` exposed on that markup extension. This should work and do what you want:

```xml
BindingContext="{inf:DependencyService Type={x:Type vm:DataBindingViewModel}, FetchTarget=NewInstance}"
```

**Comment by stezma** [comment]
Thank you so much :)

**Comment by talk2stephan** [comment]
What is the difference between adding the data context like this 
` <ContentPage.BindingContext>
        <viewModels:MainViewModel
    </ContentPage.BindingContext>`

In the above code also it will create instance of MainViewModel at every time right?

Rather write this entire thing you mentioned above

`  BindingContext="{infrastructure:DependencyService Type={x:Type viewModels:MainViewModel},
                                                      FetchTarget=NewInstance}">`

**Comment by markjulmar** [comment]
The difference is that one creates the object directly, the other goes through the dependency service - so it could be an interface/abstraction. But otherwise, you are correct - they would both create an object each time the page is created.
